File: app/services/gmail_service.py
import base64
from email import message_from_bytes
from bs4 import BeautifulSoup

from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build


# 🔹 CREATE GMAIL SERVICE USING ACCESS TOKEN
def get_gmail_service(access_token):
    creds = Credentials(token=access_token)
    return build("gmail", "v1", credentials=creds)


# 🔹 CLEAN HTML → TEXT
def clean_email_body(body):
    try:
        soup = BeautifulSoup(body, "html.parser")
        text = soup.get_text(separator=" ")
        return " ".join(text.split())
    except:
        return body


# 🔹 FILTER USEFUL EMAILS
def is_useful_email(text):
    text = text.lower()

    # allow important emails
    if any(word in text for word in ["meeting", "deadline", "submit", "project"]):
        return True

    # block obvious spam
    if "unsubscribe" in text and "offer" in text:
        return False

    return True


# 🔹 EXTRACT BODY FROM EMAIL
def extract_body(email_msg):
    text_parts = []

    if email_msg.is_multipart():
        for part in email_msg.walk():
            content_type = part.get_content_type()
            content_disposition = str(part.get("Content-Disposition"))

            if "attachment" in content_disposition:
                continue

            try:
                payload = part.get_payload(decode=True)
                if not payload:
                    continue

                decoded = payload.decode(errors="ignore")

                if content_type in ["text/plain", "text/html"]:
                    text_parts.append(decoded)

            except:
                continue
    else:
        payload = email_msg.get_payload(decode=True)
        if payload:
            text_parts.append(payload.decode(errors="ignore"))

    return "\n".join(text_parts)


# 🔥 MAIN FUNCTION (FIXED)
def fetch_threads(access_token, max_threads=20):
    service = get_gmail_service(access_token)

    thread_map = {}
    next_page_token = None

    while len(thread_map) < max_threads:
        response = service.users().messages().list(
            userId="me",
            maxResults=50,
            pageToken=next_page_token,
            q="in:inbox -category:promotions"
        ).execute()

        messages = response.get("messages", [])

        for msg in messages:
            if len(thread_map) >= max_threads:
                break

            msg_data = service.users().messages().get(
                userId="me",
                id=msg["id"],
                format="raw"
            ).execute()

            thread_id = msg_data["threadId"]
            timestamp = int(msg_data["internalDate"])

            raw_msg = base64.urlsafe_b64decode(msg_data["raw"].encode("ASCII"))
            email_msg = message_from_bytes(raw_msg)

            extracted_text = extract_body(email_msg)

            if not extracted_text.strip():
                continue

            cleaned_text = clean_email_body(extracted_text)

            if not cleaned_text.strip():
                cleaned_text = extracted_text

            # 🔥 store latest message per thread
            if thread_id not in thread_map:
                thread_map[thread_id] = {
                    "text": cleaned_text,
                    "timestamp": timestamp
                }
            else:
                if timestamp > thread_map[thread_id]["timestamp"]:
                    thread_map[thread_id] = {
                        "text": cleaned_text,
                        "timestamp": timestamp
                    }

        next_page_token = response.get("nextPageToken")
        if not next_page_token:
            break

    # 🔥 FORMAT OUTPUT
    output = []
    for thread_id, thread in thread_map.items():
        output.append({
            "thread_id": thread_id,
            "messages": [thread["text"]]
        })

    return output

File: backend/src/services/gmail.service.js
import { google } from 'googleapis';
import { simpleParser } from 'mailparser';
import * as cheerio from 'cheerio';

function getGmailService(accessToken) {
  const auth = new google.auth.OAuth2();
  auth.setCredentials({ access_token: accessToken });
  return google.gmail({ version: 'v1', auth });
}

function cleanHtml(html) {
  try {
    const $ = cheerio.load(html);
    return $.text().replace(/\s+/g, ' ').trim();
  } catch {
    return html;
  }
}

export async function fetchThreads(accessToken, maxThreads = 20) {
  const gmail = getGmailService(accessToken);
  const threadMap = new Map();
  let nextPageToken = undefined;

  while (threadMap.size < maxThreads) {
    const listRes = await gmail.users.messages.list({
      userId: 'me',
      maxResults: 50,
      pageToken: nextPageToken,
      q: 'in:inbox -category:promotions',
    });

    const messages = listRes.data.messages || [];

    for (const msg of messages) {
      if (threadMap.size >= maxThreads) break;

      const msgData = await gmail.users.messages.get({
        userId: 'me',
        id: msg.id,
        format: 'raw',
      });

      const threadId = msgData.data.threadId;
      const timestamp = parseInt(msgData.data.internalDate, 10);
      const raw = Buffer.from(msgData.data.raw, 'base64url');
      const parsed = await simpleParser(raw);

      let text = parsed.text?.trim() || '';
      if (!text && parsed.html) text = cleanHtml(parsed.html);
      if (!text) continue;

      const existing = threadMap.get(threadId);
      if (!existing || timestamp > existing.timestamp) {
        threadMap.set(threadId, { text, timestamp });
      }
    }

    nextPageToken = listRes.data.nextPageToken;
    if (!nextPageToken) break;
  }

  return Array.from(threadMap.entries()).map(([thread_id, { text }]) => ({
    thread_id,
    messages: [text],
  }));
}
